File: cogops/agents/input_guard.py
"""
cogops/agents/input_guard.py

Layer 0 — InputGuard.

Pure-code input validation: zero LLM latency. Rejects obviously malformed
input before any agent is touched.

All thresholds are configurable via config.yml under `input_guard:`.
"""

from __future__ import annotations

import math
import re
import unicodedata
from dataclasses import dataclass
from typing import Optional, Tuple


@dataclass
class GuardConfig:
    max_chars: int = 4096
    max_tokens: int = 1024
    max_repeated_char_run: int = 100
    control_char_threshold: float = 0.10
    entropy_threshold: float = 1.5
    entropy_min_length: int = 50
    token_bomb_min_words: int = 12
    token_bomb_max_avg_len: float = 2.0


# Reasons returned (used by debug telemetry; never shown to the user).
REASON_EMPTY = "empty"
REASON_TOO_LONG = "too_long"
REASON_BINARY_OR_CONTROL = "binary_or_control"
REASON_INJECTION = "injection_attempt"
REASON_SPAM = "spam"
REASON_LOW_ENTROPY = "low_entropy"
REASON_TOKEN_BOMB = "token_bomb"

# Obvious prompt-injection markers. Narrow on purpose — the real defense is
# XML data-framing in the composer prompt.
_INJECTION_PATTERNS = [
    re.compile(r"(?i)ignore (?:all |previous |above |prior )?(?:instructions|prompts|rules|directives)"),
    re.compile(r"(?i)disregard (?:all |previous |the )?(?:instructions|prompts|rules|system)"),
    re.compile(r"(?i)(?:^|\s)system\s*:\s*"),
    re.compile(r"</?(?:context|system|user|assistant|im_start|im_end)\s*>", re.IGNORECASE),
    re.compile(r"\{\{[^}]*system[^}]*\}\}"),
    re.compile(r"<\|[^|]{0,40}\|>"),
    re.compile(r"(?i)ager (?:shob |sob )?kotha (?:vule|bhule) jao"),
    re.compile(r"(?i)jailbreak|DAN mode|you are now|new instructions"),
]

_REPETITION_RE = re.compile(r"(.)\1{99,}", re.DOTALL)


class InputGuard:
    """Layer 0 — reject malformed input before any LLM call."""

    def __init__(self, config: Optional[GuardConfig] = None):
        self.cfg = config or GuardConfig()

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------
    @staticmethod
    def _control_char_fraction(text: str) -> float:
        if not text:
            return 0.0
        bad = 0
        for ch in text:
            if ch in ("\n", "\t", "\r"):
                continue
            if unicodedata.category(ch) == "Cc":
                bad += 1
        return bad / len(text)

    @staticmethod
    def _shannon_entropy(text: str) -> float:
        if not text:
            return 0.0
        freq: dict[str, int] = {}
        for ch in text:
            freq[ch] = freq.get(ch, 0) + 1
        entropy = 0.0
        length = len(text)
        for count in freq.values():
            p = count / length
            entropy -= p * math.log2(p)
        return entropy

    def _is_token_bomb(self, text: str) -> bool:
        words = [w for w in text.split() if w.strip()]
        if len(words) < self.cfg.token_bomb_min_words:
            return False
        avg_len = sum(len(w) for w in words) / len(words)
        return avg_len < self.cfg.token_bomb_max_avg_len

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------
    def check(self, query: str) -> Tuple[str, Optional[str]]:
        """Validate and normalize a user query.

        Returns:
            (clean_query, None)         on success
            ("", reason_code)           on rejection
        """
        if query is None:
            return "", REASON_EMPTY
        if not isinstance(query, str):
            return "", REASON_BINARY_OR_CONTROL

        # 1. Empty / whitespace-only
        if not query.strip():
            return "", REASON_EMPTY

        # 2. Length cap (chars, not bytes)
        if len(query) > self.cfg.max_chars:
            return "", REASON_TOO_LONG

        # 3. NUL bytes / binary
        if "\x00" in query:
            return "", REASON_BINARY_OR_CONTROL
        if self._control_char_fraction(query) > self.cfg.control_char_threshold:
            return "", REASON_BINARY_OR_CONTROL

        # 4. Injection markers
        for pat in _INJECTION_PATTERNS:
            if pat.search(query):
                return "", REASON_INJECTION

        # 5. Repetition spam
        if _REPETITION_RE.search(query):
            return "", REASON_SPAM

        # 6. Entropy check (long inputs only)
        if len(query) >= self.cfg.entropy_min_length:
            if self._shannon_entropy(query) < self.cfg.entropy_threshold:
                return "", REASON_LOW_ENTROPY

        # 7. Token bomb
        if self._is_token_bomb(query):
            return "", REASON_TOKEN_BOMB

        # All checks passed — NFC-normalize, trim, collapse whitespace.
        clean = unicodedata.normalize("NFC", query).strip()
        clean = re.sub(r"[ \t]+", " ", clean)
        clean = re.sub(r"\n{3,}", "\n\n", clean)
        return clean, None

File: mastra/src/input-guard.ts
/**
 * input-guard.ts — port of cogops/agents/input_guard.py (Layer 0).
 *
 * Pure-code validation, zero LLM latency. Returns [cleanQuery, null] on success
 * or ["", reasonCode] on rejection.
 */

export const REASON_EMPTY = "empty";
export const REASON_TOO_LONG = "too_long";
export const REASON_BINARY_OR_CONTROL = "binary_or_control";
export const REASON_INJECTION = "injection_attempt";
export const REASON_SPAM = "spam";
export const REASON_LOW_ENTROPY = "low_entropy";
export const REASON_TOKEN_BOMB = "token_bomb";

export interface GuardConfig {
  maxChars: number;
  controlCharThreshold: number;
  entropyThreshold: number;
  entropyMinLength: number;
  tokenBombMinWords: number;
  tokenBombMaxAvgLen: number;
}

const DEFAULT_CONFIG: GuardConfig = {
  maxChars: 4096,
  controlCharThreshold: 0.1,
  entropyThreshold: 1.5,
  entropyMinLength: 50,
  tokenBombMinWords: 12,
  tokenBombMaxAvgLen: 2.0,
};

const INJECTION_PATTERNS: RegExp[] = [
  /ignore (?:all |previous |above |prior )?(?:instructions|prompts|rules|directives)/i,
  /disregard (?:all |previous |the )?(?:instructions|prompts|rules|system)/i,
  /(?:^|\s)system\s*:\s*/i,
  /<\/?(?:context|system|user|assistant|im_start|im_end)\s*>/i,
  /\{\{[^}]*system[^}]*\}\}/,
  /<\|[^|]{0,40}\|>/,
  /ager (?:shob |sob )?kotha (?:vule|bhule) jao/i,
  /jailbreak|DAN mode|you are now|new instructions/i,
];

// Match any character repeated 100+ times (equiv. to Python (.)\1{99,}).
const REPETITION_RE = /(.)\1{99,}/s;

function isControlChar(cp: number): boolean {
  // Unicode category Cc: C0 (0x00-0x1F) and C1 (0x7F-0x9F).
  return (cp >= 0x00 && cp <= 0x1f) || (cp >= 0x7f && cp <= 0x9f);
}

function controlCharFraction(text: string): number {
  if (!text) return 0;
  let bad = 0;
  for (const ch of text) {
    if (ch === "\n" || ch === "\t" || ch === "\r") continue;
    const cp = ch.codePointAt(0)!;
    if (isControlChar(cp)) bad += 1;
  }
  return bad / text.length;
}

function shannonEntropy(text: string): number {
  if (!text) return 0;
  const freq = new Map<string, number>();
  for (const ch of text) freq.set(ch, (freq.get(ch) ?? 0) + 1);
  let entropy = 0;
  const length = text.length;
  for (const count of freq.values()) {
    const p = count / length;
    entropy -= p * Math.log2(p);
  }
  return entropy;
}

function isTokenBomb(text: string, cfg: GuardConfig): boolean {
  const words = text.split(/\s+/).filter((w) => w.trim());
  if (words.length < cfg.tokenBombMinWords) return false;
  const avgLen = words.reduce((s, w) => s + w.length, 0) / words.length;
  return avgLen < cfg.tokenBombMaxAvgLen;
}

export function inputGuardCheck(
  query: string | null | undefined,
  cfg: GuardConfig = DEFAULT_CONFIG,
): [string, string | null] {
  if (query === null || query === undefined) return ["", REASON_EMPTY];
  if (typeof query !== "string") return ["", REASON_BINARY_OR_CONTROL];
  if (!query.trim()) return ["", REASON_EMPTY];
  if (query.length > cfg.maxChars) return ["", REASON_TOO_LONG];
  if (query.includes("\x00")) return ["", REASON_BINARY_OR_CONTROL];
  if (controlCharFraction(query) > cfg.controlCharThreshold) return ["", REASON_BINARY_OR_CONTROL];
  for (const pat of INJECTION_PATTERNS) {
    if (pat.test(query)) return ["", REASON_INJECTION];
  }
  if (REPETITION_RE.test(query)) return ["", REASON_SPAM];
  if (query.length >= cfg.entropyMinLength && shannonEntropy(query) < cfg.entropyThreshold) {
    return ["", REASON_LOW_ENTROPY];
  }
  if (isTokenBomb(query, cfg)) return ["", REASON_TOKEN_BOMB];

  // NFC-normalize, trim, collapse whitespace.
  let clean = query.normalize("NFC").trim();
  clean = clean.replace(/[ \t]+/g, " ");
  clean = clean.replace(/\n{3,}/g, "\n\n");
  return [clean, null];
}
